### DISCUSSION TITLE
GKE DNS Endpoint Support

### DISCUSSION THREAD

**Discussion by jjtroberts** [argoproj/argo-cd#20640] [General]
There is a new [DNS endpoint feature](https://cloud.google.com/kubernetes-engine/docs/concepts/network-isolation#dns-based_endpoint) for GKE kube-apiserver that, when enabled, creates a DNS endpoint similar to gke-some-long-hash.region.gke.goog.

This endpoint is protected by IAM, meaning that any request must be able to authenticate as a GCP IAM identity and has container.clusters.connect which is included in any identity that has either of the roles:
- roles/container.developer
- roles/container.viewer

We had Workload Identity setup and working as expected prior to this change.

I enabled the dns endpoint in a lab environment and updated my ArgoCD cluster secret and project destination but am unable to connect:
```
Get "https://gke-xxxxxxx.us-east4.gke.goog/version?timeout=32s": net/http: TLS handshake timeout - error from a previous attempt: read tcp xxx.xxx.xxx.xxx:52834->xxx.xxx.xxx.xxx:443: read: connection reset by peer
```

I've tried with tlsClientConfig.insecure set to false and to true, and without a tlsClientConfig at all. Any insight into how ArgoCD authenticates using the execProviderConfig below would be appreciated.

```
config: |
    {
      "execProviderConfig": {
        "command": "argocd-k8s-auth",
        "args": ["gcp"],
        "apiVersion": "client.authentication.k8s.io/v1beta1"
      },
      "tlsClientConfig": {
        "insecure": true
      }
    }
```

**Comment by rekiemfaxaf** [comment]
Hey, do you found any way to make it work? I'm wanting to test the same scenario.
Greetings!

**Reply by jjtroberts** [reply]
We ended up using Cloud NAT in our ArgoCD project, and then added the NAT egress IP addresses to the GKE cluster's master authorized networks. So we're effectively using both public and private endpoints. Public with ip allow listing for services, private IAM for users.

**Comment by rekiemfaxaf** [comment]
Is there any update regarding this? Will be great to see this working on version 3.x

**Reply by jjtroberts** [reply]
We eventually got this working. I cannot say which release it was introduced in, or if it always worked an we just missed something. But here's what worked for us using v2.13.2. One of the keys was to leave `caData` empty.

```
apiVersion: v1
kind: Secret
metadata:
  name: gke-cluster-secret
  namespace: argocd
  labels:
    argocd.argoproj.io/secret-type: cluster
type: Opaque
stringData:
  name: gke-cluster-secret
  server: https://gke-some-really-long-guid.us-central1.gke.goog
  config: |
    {
      "execProviderConfig": {
        "command": "argocd-k8s-auth",
        "args": ["gcp"],
        "apiVersion": "client.authentication.k8s.io/v1beta1"
      },
      "tlsClientConfig": {
        "insecure": false,
        "caData": ""
      }
    }
```

**Reply by mpenalozag** [reply]
@jjtroberts hey! Awesome to see you got this to work. I'm currently trying to do the same on my cluster, having one argoCD instance that manages my other clusters. However it's not working for me. I'm not sure what I'm missing. I setup workload identity so the KSAs from argoCD use the IAM permissions of my service account that has access to the target cluster I'm trying to manage. Have tried different configs for the secret that contains the cluster auth data but still nothing. Any tips on what I could pay attention to? I'm sure I must be missing a small step.

PS: Just got it to work! Not sure why removing the `caData` solves this though

**Reply by mpenalozag** [reply]
Btw, the reason you have to remove the caData is because DNS-endpoint uses Google Trust Services for the certificate, which is a public certificate authority and likely lives on your default CA Trust Store
